false-positive(checker): tail never-call expression statement not treated as terminating in return-type inference -> body inferred `void` not `never` (TS2322 vs () => boolean) (immer)

## Summary
A function/method whose body is a single tail expression statement calling a never-returning function (e.g. `die(n): never`, immer's error helper) is inferred to return `void` instead of `never`. When that function is contextually typed or annotated as `() => boolean`, tsz emits `TS2322 Type '() => void' is not assignable ... Type 'void' is not assignable to type 'boolean'`. tsc 5.9.3 accepts it because the body's endpoint is unreachable (no implicit `return undefined`). `throw` is handled correctly; only a never-RETURNING CALL in tail position is missed. In **immer** this fires at `src/core/proxy.ts(241)` `defineProperty() { die(11) }` and `(247)` `setPrototypeOf() { die(12) }` (2 of the 9 tsz-only FPs).

## Repro
```ts
declare function die(n: number): never

interface Handler {
  defineProperty: (target: object, k: string | symbol, d: PropertyDescriptor) => boolean
}

const h: Handler = {
  defineProperty() {
    die(11)
  }
}
```
- tsz (current origin/main): `(8,3): error TS2322: Type '() => void' is not assignable to type '(target: object, k: string | symbol, d: PropertyDescriptor) => boolean'.` / `Type 'void' is not assignable to type 'boolean'.`
- tsc 5.9.3: clean (exit 0)

Controls that isolate the cause:
- Replace `die(11)` with `throw new Error()` -> CLEAN in tsz (throw recognized as terminating).
- Arrow form `const fn: () => boolean = () => { die(1) }` -> same TS2322.
- Standalone `function f() { die(11) }` with NO contextual type/annotation -> infers `void` in tsc TOO, so that case is correct and not part of this bug; the divergence is exactly the contextually-typed / annotated path.

## Root cause (file:line)
Wrong operation: reachability-driven return-type inference. In `crates/tsz-checker/src/types/utilities/return_type.rs`:
- `body_contains_throw_or_never_call` (lines 152-220) only scans for `THROW_STATEMENT` (and recurses into blocks/if/try/switch); it returns false for `EXPRESSION_STATEMENT` (explicit comment lines 202-205: 'the full falls_through check will catch them').
- `infer_return_type_from_body_inner` line 1070: `return if !may_not_fall_through || self.function_body_falls_through(body_idx) { ...VOID... } else { NEVER }`. For body `{ die(11) }`, `may_not_fall_through = body_contains_throw_or_never_call = false`, so `!false = true` short-circuits the `||` and `function_body_falls_through` is never called. The never-branch at lines 1085-1086 is unreachable for tail never-calls.

`function_body_falls_through` -> `block_falls_through` (crates/tsz-checker/src/flow/reachability_checker.rs) DOES recognize never-calls via `call_expression_terminates_control_flow` -> `callee_explicitly_returns_never` -> `declaration_explicitly_returns_never` (reads the `: never` annotation, lines 247-259), so the capability exists; it is just gated out by the short-circuit.

## Path to fixing
- Make the gate consult `function_body_falls_through` whenever the body has no return statements, regardless of `may_not_fall_through`; or extend `body_contains_throw_or_never_call` to detect a tail EXPRESSION_STATEMENT whose call terminates control flow (reuse `call_expression_terminates_control_flow`). Either way, a no-return body whose endpoint is unreachable via a never-call must infer `never`, not `void`.
- Guard against the documented recursion concern (named function expression calling itself) by only checking the LAST statement / using the existing `check_body_for_throws=false` symbol path that reads annotations rather than re-analyzing bodies.
- Tests: contextual arrow + method-shorthand + standalone-with-annotation forms; `die`-style annotated `never` function, imported `die` (alias path already in `symbol_explicitly_returns_never`), and a `throw` control; vary the binder name to avoid name fast paths.

## Adjacent cases
- Method shorthand and arrow both reproduce (immer uses method shorthand in the ProxyHandler object literal).
- Imported never-function (`import { die }`) — the reachability checker already has an alias path (`symbol_explicitly_returns_never`, reachability_checker.rs:235-238); ensure the inference gate reaches it.
- Annotated-return form `function f(): boolean { die(11) }` should be accepted (no TS2366 'lacks ending return') — verify the same fix covers `has_return_type_annotation` branch at lines 1036-1047.
- `process.exit()`-style never calls and assertion calls (`call_expression_terminates_control_flow` already handles assertion-false) should behave consistently.

Goal: green

## Provenance
Machine: m1 · Assistant: claude-code · Model: claude-opus-4-8 · M1FableArchitect

--- a/crates/tsz-checker/src/types/utilities/return_type.rs
+++ b/crates/tsz-checker/src/types/utilities/return_type.rs
@@ -147,9 +147,13 @@ impl<'a> CheckerState<'a> {
     /// }
     /// ```
     /// Lightweight AST scan: does the function body contain any `throw` statements?
-    /// This is used as a pre-check before the more expensive `function_body_falls_through`
-    /// to avoid triggering type evaluation in simple function bodies that obviously fall through.
-    fn body_contains_throw_or_never_call(&self, body_idx: NodeIndex) -> bool {
+    ///
+    /// This is a syntax-only pre-check; it deliberately does not detect
+    /// never-returning tail calls (e.g. `die(11)`), which require resolving the
+    /// callee's signature. The authoritative reachability answer for the
+    /// `void` vs `never` inference is `function_body_falls_through`, which routes
+    /// expression-statement calls through `call_expression_terminates_control_flow`.
+    fn body_contains_throw(&self, body_idx: NodeIndex) -> bool {
         fn scan_stmts(arena: &tsz_parser::parser::NodeArena, stmts: &[NodeIndex]) -> bool {
             use tsz_parser::parser::syntax_kind_ext;
             for &idx in stmts {
@@ -1019,10 +1023,14 @@ impl<'a> CheckerState<'a> {
             // No return statements found. Check if the body falls through:
             // - If it does (normal implicit return), the return type is `void`
             // - If it doesn't (all paths throw or call never), the return type is `never`
-            // Only call the (potentially expensive) fallthrough checker when the body
-            // could plausibly be non-falling-through, i.e. it contains throw statements.
-            // This avoids triggering unnecessary type evaluation in simple function bodies.
-            let may_not_fall_through = self.body_contains_throw_or_never_call(body_idx);
+            // `function_body_falls_through` is the authoritative reachability answer;
+            // it routes expression-statement calls through
+            // `call_expression_terminates_control_flow`, so a tail never-returning
+            // call (`die(11)`) terminates the body and infers `never`, not `void`.
+            // The throw-only syntax pre-scan never observed those tail calls, which
+            // wrongly short-circuited contextually-typed/inferred bodies to `void`
+            // (#14741). Compute reachability once and reuse it below.
+            let falls_through = self.function_body_falls_through(body_idx);
 
             // Check if function has a return type annotation
             let has_return_type_annotation = if let Some(func_node) = self.ctx.arena.get(body_idx)
@@ -1033,10 +1041,7 @@ impl<'a> CheckerState<'a> {
                 false
             };
 
-            if has_return_type_annotation
-                && may_not_fall_through
-                && !self.function_body_falls_through(body_idx)
-            {
+            if has_return_type_annotation && !falls_through && self.body_contains_throw(body_idx) {
                 use crate::diagnostics::diagnostic_codes;
                 self.error_at_node(
                     body_idx,
@@ -1067,7 +1072,7 @@ impl<'a> CheckerState<'a> {
                 }
             }
 
-            return if !may_not_fall_through || self.function_body_falls_through(body_idx) {
+            return if falls_through {
                 // When contextual return type expects `undefined` (not void/any/unknown),
                 // use `undefined` so `const f: () => undefined = () => {}` doesn't produce TS2322.
                 // tsc applies contextual typing to infer the return type of such lambdas.

--- a/crates/tsz-checker/src/types/utilities/return_type.rs
+++ b/crates/tsz-checker/src/types/utilities/return_type.rs
@@ -223,6 +223,61 @@ impl<'a> CheckerState<'a> {
         false
     }
 
+    /// Lightweight, syntax-only check: can the body's terminal statement plausibly
+    /// fail to fall through (e.g. a never-returning tail call `die(11)`, a
+    /// structured `if`/`switch`/`try`, or a jump statement)?
+    ///
+    /// This gates the (potentially evaluation-triggering)
+    /// `function_body_falls_through` query. When the body's last statement is a
+    /// trivially-falling-through statement (a variable/function/class declaration,
+    /// a non-call expression statement, etc.), the body provably falls through, so
+    /// the return type is `void` and we must NOT run the reachability query —
+    /// doing so would evaluate flow-sensitive receivers (e.g. evolving
+    /// implicit-`any` arrays) prematurely and suppress their `TS7005`/`TS7034`
+    /// diagnostics. A never-returning tail call always appears as the terminal
+    /// expression statement, so it is preserved here (#14741).
+    fn body_tail_may_terminate_control_flow(&self, body_idx: NodeIndex) -> bool {
+        let Some(body_node) = self.ctx.arena.get(body_idx) else {
+            return false;
+        };
+        if body_node.kind != syntax_kind_ext::BLOCK {
+            return false;
+        }
+        let Some(block) = self.ctx.arena.get_block(body_node) else {
+            return false;
+        };
+        let Some(&last_idx) = block.statements.nodes.last() else {
+            return false;
+        };
+        let Some(last) = self.ctx.arena.get(last_idx) else {
+            return false;
+        };
+        match last.kind {
+            syntax_kind_ext::THROW_STATEMENT
+            | syntax_kind_ext::RETURN_STATEMENT
+            | syntax_kind_ext::BREAK_STATEMENT
+            | syntax_kind_ext::CONTINUE_STATEMENT
+            | syntax_kind_ext::IF_STATEMENT
+            | syntax_kind_ext::SWITCH_STATEMENT
+            | syntax_kind_ext::TRY_STATEMENT
+            | syntax_kind_ext::BLOCK
+            | syntax_kind_ext::LABELED_STATEMENT
+            | syntax_kind_ext::WHILE_STATEMENT
+            | syntax_kind_ext::DO_STATEMENT
+            | syntax_kind_ext::FOR_STATEMENT => true,
+            syntax_kind_ext::EXPRESSION_STATEMENT => self
+                .ctx
+                .arena
+                .get_expression_statement(last)
+                .and_then(|stmt| self.ctx.arena.get(stmt.expression))
+                .is_some_and(|expr| {
+                    expr.kind == syntax_kind_ext::CALL_EXPRESSION
+                        || expr.kind == syntax_kind_ext::NEW_EXPRESSION
+                }),
+            _ => false,
+        }
+    }
+
     pub fn function_body_falls_through(&mut self, body_idx: NodeIndex) -> bool {
         let Some(body_node) = self.ctx.arena.get(body_idx) else {
             return true;
@@ -1059,14 +1114,20 @@ impl<'a> CheckerState<'a> {
             // No return statements found. Check if the body falls through:
             // - If it does (normal implicit return), the return type is `void`
             // - If it doesn't (all paths throw or call never), the return type is `never`
-            // `function_body_falls_through` is the authoritative reachability answer;
-            // it routes expression-statement calls through
+            // `function_body_falls_through` is the authoritative reachability
+            // answer; it routes expression-statement calls through
             // `call_expression_terminates_control_flow`, so a tail never-returning
-            // call (`die(11)`) terminates the body and infers `never`, not `void`.
-            // The throw-only syntax pre-scan never observed those tail calls, which
-            // wrongly short-circuited contextually-typed/inferred bodies to `void`
-            // (#14741). Compute reachability once and reuse it below.
-            let falls_through = self.function_body_falls_through(body_idx);
+            // call (`die(11)`) terminates the body and infers `never`, not `void`
+            // (#14741). But that query can evaluate flow-sensitive receivers (e.g.
+            // `x.push(...)` on an evolving implicit-`any` array), so running it on
+            // bodies that obviously fall through prematurely freezes those types
+            // and drops `TS7005`/`TS7034`. Gate it behind a cheap syntax pre-check:
+            // a `throw` anywhere, or a terminal statement that can plausibly
+            // terminate control flow (the never-returning tail call lives there).
+            // Otherwise the body provably falls through and returns `void`.
+            let may_not_fall_through = self.body_contains_throw(body_idx)
+                || self.body_tail_may_terminate_control_flow(body_idx);
+            let falls_through = !may_not_fall_through || self.function_body_falls_through(body_idx);
 
             // Check if function has a return type annotation
             let has_return_type_annotation = if let Some(func_node) = self.ctx.arena.get(body_idx)
@@ -1108,7 +1169,11 @@ impl<'a> CheckerState<'a> {
                 }
             }
 
-            return if falls_through {
+            // A bare `return;` (`saw_empty`) means the body explicitly returns
+            // `void`/`undefined`, never `never` — even though it does not fall off
+            // the end. Only bodies with no `return` whose every path throws or
+            // calls a never-returning function infer `never`.
+            return if falls_through || saw_empty {
                 // When contextual return type expects `undefined` (not void/any/unknown),
                 // use `undefined` so `const f: () => undefined = () => {}` doesn't produce TS2322.
                 // tsc applies contextual typing to infer the return type of such lambdas.

--- a/crates/tsz-checker/tests/function_terminal_return_flow_tests.rs
+++ b/crates/tsz-checker/tests/function_terminal_return_flow_tests.rs
@@ -92,3 +92,126 @@ function renamedUnknown(flag: boolean): unknown {
         "terminal bare return should not look like an empty falling-through unknown body; got {codes:?}"
     );
 }
+
+// Regression for #14741: a no-return body whose endpoint is unreachable via a
+// tail call to a `never`-returning function must infer `never`, not `void`, in
+// the contextually-typed / inferred path. Previously the throw-only syntax
+// pre-scan short-circuited the reachability check, so these bodies inferred
+// `void` and produced a spurious TS2322 against the contextual signature. All
+// binder names are renamed so no fixture-name fast path can satisfy the case.
+
+#[test]
+fn tail_never_call_in_contextual_method_shorthand_infers_never() {
+    let codes = diagnostic_codes(
+        r#"
+declare function bailOut(code: number): never
+
+interface RenamedHandler {
+    defineProperty: (target: object, key: string | symbol, desc: PropertyDescriptor) => boolean
+}
+
+const renamedHandler: RenamedHandler = {
+    defineProperty() {
+        bailOut(11)
+    }
+}
+"#,
+        false,
+    );
+
+    assert!(
+        !codes.contains(&2322),
+        "tail never-call in a contextually-typed method must infer never, not void; got {codes:?}"
+    );
+}
+
+#[test]
+fn tail_never_call_in_contextual_arrow_infers_never() {
+    let codes = diagnostic_codes(
+        r#"
+declare function abortNow(code: number): never
+const renamedArrow: () => boolean = () => { abortNow(1) }
+"#,
+        false,
+    );
+
+    assert!(
+        !codes.contains(&2322),
+        "tail never-call in a contextually-typed arrow must infer never, not void; got {codes:?}"
+    );
+}
+
+#[test]
+fn tail_namespace_never_call_in_contextual_arrow_infers_never() {
+    let codes = diagnostic_codes(
+        r#"
+namespace RenamedDebug {
+    export function fail(): never { throw new Error() }
+}
+const renamedNs: () => string = () => { RenamedDebug.fail() }
+"#,
+        false,
+    );
+
+    assert!(
+        !codes.contains(&2322),
+        "tail namespace-qualified never-call must infer never, not void; got {codes:?}"
+    );
+}
+
+#[test]
+fn tail_this_method_never_call_in_contextual_arrow_infers_never() {
+    let codes = diagnostic_codes(
+        r#"
+interface RenamedBox { compute: () => number }
+class RenamedComputer {
+    panic(): never { throw new Error() }
+    make(): RenamedBox {
+        return { compute: () => { this.panic() } }
+    }
+}
+"#,
+        false,
+    );
+
+    assert!(
+        !codes.contains(&2322),
+        "tail this.method never-call must infer never, not void; got {codes:?}"
+    );
+}
+
+#[test]
+fn tail_assertion_false_call_in_contextual_arrow_infers_never() {
+    let codes = diagnostic_codes(
+        r#"
+declare function renamedAssert(cond: boolean): asserts cond
+const renamedAssertArrow: () => number = () => { renamedAssert(false) }
+"#,
+        false,
+    );
+
+    assert!(
+        !codes.contains(&2322),
+        "tail assertion call with a false condition terminates control flow; got {codes:?}"
+    );
+}
+
+#[test]
+fn tail_void_call_in_contextual_arrow_still_reports_ts2322() {
+    // Negative control: a real `void`-returning tail call does NOT terminate
+    // control flow, so the body still infers `void` and the contextual
+    // `() => boolean` mismatch must still surface (parity with tsc). This guards
+    // against over-applying the `never` inference.
+    let codes = diagnostic_codes(
+        r#"
+declare function renamedEffect(): void
+const renamedVoidArrow: () => boolean = () => { renamedEffect() }
+"#,
+        false,
+    );
+
+    assert!(
+        codes.contains(&2322),
+        "a real void tail body must still mismatch the contextual boolean return; got {codes:?}"
+    );
+}

